### PR TITLE
Pcap Directory Mode

### PR DESCRIPTION
**Pull request by dbcfd**
Version 5 of 
 - https://github.com/OISF/suricata/pull/2961
 - https://github.com/OISF/suricata/pull/2960
 - https://github.com/OISF/suricata/pull/2958
 - https://github.com/OISF/suricata/pull/2957

Pcap file mode that when passed a directory will process all files in that directory until the directory is moved/deleted or suricata is interrupted.

Engine state will not reset between files.

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [X] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [X] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [X] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2222

Describe changes:
- If source-pcap-file is passed a directory instead of a file, all files in that directory will be processed in order of modified time, without state resetting between files.
- source-pcap-file will continue to process files until suricata is interrupted or the directory is moved/deleted
- runmode-unix-socket is updated to provide a current time to determine what state of processing of the directory is

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

### CODE REVIEW

File: src/source-pcap-file.c [hunk 1]
```diff
@@ -45,6 +45,14 @@
 #include "util-checksum.h"
 #include "util-atomic.h"
 
+//directory support
+#include <dirent.h> // dirent
```
Comment (victorjulien) [inline src/source-pcap-file.c]: these should go into suricata-common.h, properly wrapped in ifdefs and checked for in configure.ac

Comment (victorjulien) [reply src/source-pcap-file.c]: most of them should already be there

File: src/source-pcap-file.c [hunk 2]
```diff
@@ -60,38 +68,76 @@
...
+    char *filename;
+    DIR *directory;
+    PcapFileFileVars *current_file;
+    time_t delay;
+    time_t poll_interval;
+
+    PcapFileSharedVars *shared;
+} PcapFileDirectoryVars;
+
+typedef union PcapFileBehaviorVar_
```
Comment (victorjulien) [inline src/source-pcap-file.c]: need a comment as it's not self explaining

File: src/source-pcap-file.c [hunk 3]
```diff
@@ -60,38 +68,76 @@
...
+{
+    PcapFileDirectoryVars *directory;
+    PcapFileFileVars *file;
+} PcapFileBehaviorVar;
+
+typedef struct PcapFileThreadVars_
+{
+    PcapFileBehaviorVar behavior;
+    int is_directory;
```
Comment (victorjulien) [inline src/source-pcap-file.c]: minor: 'bool'

File: src/source-pcap-file.c [hunk 4]
```diff
@@ -101,6 +147,351 @@ TmEcode DecodePcapFile(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueu
...
+void CleanupPcapFileDirectoryVars(PcapFileThreadVars *tv, PcapFileDirectoryVars *ptv);
+void CleanupPcapFileFileVars(PcapFileThreadVars *tv, PcapFileFileVars *ptv);
+void CleanupPcapFileThreadVars(PcapFileThreadVars *tv);
+TmEcode PcapDirectoryFailure(PcapFileThreadVars *tv, PcapFileDirectoryVars *ptv);
+TmEcode PcapDirectoryDone(PcapFileThreadVars *tv, PcapFileDirectoryVars *ptv);
+TmEcode PcapCheckFile(char *filename, DIR **directory);
+int PcapDirectoryGetModifiedTime(char const * file, struct timespec * out);
+int PcapDirectorySortByStatTime(const void * vleft, const void * vright);
+void FreeDirectoryMemBuffer(MemBuffer * buffer);
+TmEcode PcapDirectoryPopulateBuffer(PcapFileDirectoryVars *ptv, time_t newer_than, time_t older_than, time_t *first_new_file_time, MemBuffer **directory_content);
```
Comment (victorjulien) [inline src/source-pcap-file.c]: wrap lines

File: src/source-pcap-file.c [hunk 5]
```diff
@@ -101,6 +147,351 @@ TmEcode DecodePcapFile(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueu
...
+                break;
+
+            case ENOENT:
+                SCLogError(SC_ERR_FOPEN, "%s: Does not exist, or name is an empty string", filename);
+                break;
+            case ENOMEM:
+                SCLogError(SC_ERR_FOPEN, "%s: Insufficient memory to complete the operation", filename);
+                break;
+
+            case ENOTDIR:
```
Comment (victorjulien) [inline src/source-pcap-file.c]: can you add a comment to explain why this is handled differently?

File: src/source-pcap-file.c [hunk 6]
```diff
@@ -101,6 +147,351 @@ TmEcode DecodePcapFile(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueu
...
+
+    return ret;
+}
+
+int PcapDirectorySortByStatTime(const void * vleft, const void * vright)
+{
+    if (vleft == vright)
+        return 0;
+
+    struct timespec leftTime, rightTime;
```
Comment (victorjulien) [inline src/source-pcap-file.c]: coding style: we use snake_case for vars

File: src/source-pcap-file.c [hunk 7]
```diff
@@ -101,6 +147,351 @@ TmEcode DecodePcapFile(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueu
...
+{
+    if (buffer == NULL)
+        return;
+
+    size_t offset = 0;
+    char *buffer_value;
+
+    while (offset < MEMBUFFER_OFFSET(buffer))
+    {
+        buffer_value = *((char**)(MEMBUFFER_BUFFER(buffer) + offset));
```
Comment (victorjulien) [inline src/source-pcap-file.c]: this looks suspicious to me, why is a single call to MemBufferFree not sufficient?

File: src/source-pcap-file.c [hunk 8]
```diff
@@ -101,6 +147,351 @@ TmEcode DecodePcapFile(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueu
...
+
+                SCLogError(SC_ERR_MEM_ALLOC, "Failed to copy file name");
+
+                SCReturnInt(TM_ECODE_FAILED);
+            }
+            memcpy(mem, pathbuff, written + 1);
+
+            SCLogDebug("Found \"%s\"", mem);
+
+            *((char**)(MEMBUFFER_BUFFER(temp_directory_content) + MEMBUFFER_OFFSET(temp_directory_content))) = mem;
```
Comment (victorjulien) [inline src/source-pcap-file.c]: this is definitely not the MemBuffer's intended use. It seems a bit over-complicated too. Why not use a simple list/TAILQ?

File: .gitignore [hunk 9]
```diff
@@ -1,3 +1,4 @@
+*.DS_Store
```
Comment (victorjulien) [inline .gitignore]: this should be in it's own commit

File: src/util-logopenfile.c [hunk 10]
```diff
@@ -65,13 +65,17 @@ SCLogOpenUnixSocketFp(const char *path, int sock_type, int log_err)
     saun.sun_family = AF_UNIX;
     strlcpy(saun.sun_path, path, sizeof(saun.sun_path));
 
+    SCLogInfo("Log connecting to unix socket %s", path);
```
Comment (victorjulien) [inline src/util-logopenfile.c]: these changes seem unrelated to pcap directory processing and should go into their own commit

File: src/suricata.c [hunk 11]
```diff
@@ -2270,9 +2270,6 @@ void PreRunPostPrivsDropInit(const int runmode)
  * Will be run once per pcap in unix-socket mode */
 void PostRunDeinit(const int runmode, struct timeval *start_time)
 {
-    if (runmode == RUNMODE_UNIX_SOCKET)
```
Comment (victorjulien) [inline src/suricata.c]: in what case will this be called with runmode == RUNMODE_UNIX_SOCKET ?

File: src/source-pcap-file.c [hunk 12]
```diff
@@ -101,6 +147,351 @@ TmEcode DecodePcapFile(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueu
...
+            FreeDirectoryMemBuffer(temp_directory_content);
+
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+    }
+
+    if (temp_directory_content != NULL)
+    {
+        // Sort buffer
+        qsort(MEMBUFFER_BUFFER(temp_directory_content), MEMBUFFER_OFFSET(temp_directory_content) / sizeof(char*), sizeof(char*), &PcapDirectorySortByStatTime);
```
Comment (victorjulien) [inline src/source-pcap-file.c]: wrap line

File: src/source-pcap-file.c [hunk 13]
```diff
@@ -101,6 +147,351 @@ TmEcode DecodePcapFile(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueu
 TmEcode DecodePcapFileThreadInit(ThreadVars *, const void *, void **);
 TmEcode DecodePcapFileThreadDeinit(ThreadVars *tv, void *data);
 
+TmEcode InitPcapFile(PcapFileFileVars *pfv, const char *filename);
```
Comment (victorjulien) [inline src/source-pcap-file.c]: all these prototypes should be 'static' so we see when they are unused

**Comment by dbcfd** [comment]
Closing for https://github.com/OISF/suricata/pull/2971
